docs: add docs for different PyPI mirror (#113)
lumicks/pylake: docs/install.rst

    pip install lumicks.pylake

> Alternatively, if you're using Anaconda, please see the more detailed instructions below.

Reviewer: ```suggestion
Alternatively, if you are using Anaconda, please see the more detailed instructions below.
```

.. rubric:: Windows

#. Go to the `Anaconda`_ website and download the Python 3.8 (or newer) installer.

#. Run it and accept the default options during the installation.

#. Open `Anaconda Prompt` from the `Start` menu.

#. Create a new environment for Pylake named pylake by invoking the following command::

    conda create -n pylake

#. The environment can then be activated::

    conda activate pylake

> #. The line where you enter your next command should now start with the text `(pylake)` instead of `(base)`, to indicate you're in the pylake environment you just created.

Reviewer: ```suggestion
#. The line where you enter your next command should now start with the text `(pylake)` instead of `(base)`, to indicate you are in the pylake environment you just created.
```

#. Pylake can be found on `conda-forge`_. Add `conda-forge`_ to the list of sources as follows::

    conda config --add channels conda-forge

> #. We can install Pylake in this environment by invoking the following commands::

Reviewer: ```suggestion
#. We can install Pylake in this environment by invoking the following command::
```

    conda install lumicks.pylake

#. To make sure Jupyter notebook can find the ipykernel, we invoke::

    python -m ipykernel install --user

#. It should be possible to open a jupyter notebook in this environment by calling::

    jupyter notebook

Note that if you are used to starting Jupyter notebook from the Anaconda Navigator, you will have to set the environment to `pylake` for it to have access to the environment where you just installed Pylake.

You can do this in the drop down menu that normally defaults to the environment `base`.

If you are used to starting Jupyter from the Anaconda prompt, then remember to activate the correct environment (pylake) prior to starting the notebook.
You can activate this environment by calling::

    conda activate pylake

> That's it, all done. Check out the :doc:`Tutorial </tutorial/index>` for some example code and Jupyter notebooks to get started.

Reviewer: ```suggestion
That is all. Check out the :doc:`Tutorial </tutorial/index>` for some example code and Jupyter notebooks to get started.
```

.. rubric:: Linux

#. Go to the `Anaconda`_ website and download the Python 3.8 (or newer) installer.

#. Open a terminal window and run::

    bash Anaconda3-x.x.x-Linux-x86_64.sh

#. Follow the installation steps. Make sure that you type `yes` when prompted with the question whether to run conda init::

    Do you wish the installer to initialize Anaconda 3 by running conda init? [yes|no]
    [no] >>> yes

#. Now, close your terminal window and open a new one for the changes to take effect.

#. Using the terminal, we will create a new environment for Pylake named pylake by invoking the following command::

    conda create -n pylake

#. The environment can then be activated::

    conda activate pylake

#. Pylake can be found on `conda-forge`_. Add `conda-forge`_ to the list of sources as follows::

    conda config --add channels conda-forge

> #. We can install Pylake in this environment by invoking the following commands::

Reviewer: ```suggestion
#. We can install Pylake in this environment by invoking the following command::
```

    conda install lumicks.pylake

> #. It should be possible to open a Jupyter notebook in this environment by calling `jupyter notebook` from the terminal.
> 
> #. It should also be possible to now start the Anaconda Navigator by calling `anaconda-navigator` from the terminal.
> 

Reviewer: ```suggestion
#. You can open a Jupyter notebook in this environment by calling `jupyter notebook` from the terminal.

#. You can also now start the Anaconda Navigator by calling `anaconda-navigator` from the terminal.

```

Note that if you are used to starting Jupyter notebook from the Anaconda Navigator, you will have to set the environment to `pylake` for it to have access to the environment where you just installed Pylake.
You can do this in the drop down menu that normally defaults to the environment `base`.
If you are used to starting Jupyter notebook from the terminal, then remember to activate the correct environment (pylake) prior to starting the notebook.

> That's it, all done. Check out the :doc:`Tutorial </tutorial/index>` for some example code and Jupyter notebooks to get started.

Reviewer: ```suggestion
That is all. Check out the :doc:`Tutorial </tutorial/index>` for some example code and Jupyter notebooks to get started.
```

.. rubric:: macOS

#. Go to the `Anaconda`_ website and download the Python 3.8 (or newer) installer.

#. Run it and accept the default options during the installation.

#. Open `Terminal`. First, we will create a new environment for Pylake named pylake by invoking the following command::

    conda create -n pylake

#. The environment can then be activated by invoking the following::

    conda activate pylake

#. Pylake can be found on `conda-forge`_. We can add `conda-forge`_ to the list of sources as follows::

    conda config --add channels conda-forge

> #. Install Pylake in this environment by invoking the following commands::

Reviewer: ```suggestion
#. Install Pylake in this environment by invoking the following command::
```

    conda install lumicks.pylake

> #. It should be possible to open a jupyter notebook in this environment by calling::

Reviewer: ```suggestion
#. You can open a jupyter notebook in this environment by calling::
```

    jupyter notebook

Note that if you are used to starting Jupyter notebook from the Anaconda Navigator, you will have to set the environment to `pylake` for it to have access to the environment where you just installed pylake.
You can do this in the drop down menu that normally defaults to the environment `base`.
If you are used to starting Jupyter from `Terminal`, then remember to activate the correct environment (pylake) prior to starting the notebook.

> That's it, all done. Check out the :doc:`Tutorial </tutorial/index>` for some example code and Jupyter notebooks to get started.

Reviewer: ```suggestion
That is all. Check out the :doc:`Tutorial </tutorial/index>` for some example code and Jupyter notebooks to get started.
```

.. _VS Code: https://code.visualstudio.com/download
.. _VS Code Environment Instructions: https://code.visualstudio.com/docs/python/environments#_conda-environments
.. _numpy discussion: https://github.com/numpy/numpy/issues/15183#issuecomment-603575874

If you have installed Pylake according to the Installation Instructions for Anaconda, then you should now have a separate environment for your Pylake work.
You may be wondering why we needed to create a new environment for Pylake, and what an environment is.

When using Python, you will quickly find that several packages that you can install depend on each other.
For example, when installing Pylake, you could see that the installation of Pylake required numerous other packages to be installed as well.
Most Python packages are continuously updated by their authors. Sometimes, these authors decide that certain existing functionalities need to be changed.
This means that not all of them will be completely backwards compatible.
Therefore, it can be challenging to find a set of packages and package versions that all work together.

One pragmatic solution to this is to maintain separate Python environments for different projects.

> This means that you create independent "copies" of Python and its installed packages, so that the different projects you're working on don't interfere with each other.

Reviewer: ```suggestion
This means that you create independent "copies" of Python and its installed packages, so that the different projects you are working on do not interfere with each other.
```

Anaconda is one solution to this problem. With Anaconda, you can have multiple installations of Python (with all their installed modules) installed on your computer.
These installations are referred to as environments.

**Why do we install Pylake in a separate environment by default?**

> Conda sources the packages it uses from a channel, these are locations where conda and the Anaconda Navigator search for packages.

Reviewer: ```suggestion
Conda fetches the packages it uses from a channel, these are locations where conda and the Anaconda Navigator search for packages.
```

Make sure that you selected the pylake environment when starting VS Code from the Anaconda Navigator.
It should now be possible to use pylake in VS Code.

Under Windows, you will need to start VS Code from the Anaconda Navigator for it to use the correct environment.
For more information on how to get it to run without Anaconda Navigator on Windows see the `VS Code Environment Instructions`_ and this `numpy discussion`_.

**Can I use pip with Anaconda?**

`Pip` is a different package manager.
While `conda` does allow you to install `pip` inside a conda environment, there is no guarantee that `pip` packages will be compatible with `conda-forge` packages.
It is therefore wise to choose one package manager as your go-to package manager and only switch when a package you need can only be found on the other package manager.

While using `pip` within conda is perfectly possible, note that if you do decide to go down this route, you should install all packages in that environment via `pip` and none via `conda`.
If you decide to use this configuration then you have to make sure that you install it using the version of `pip` inside your conda environment.

> You can get problematic incompatibilities if you use a system-wide install of `pip` in conjunction with an active `conda` environment.

Reviewer: ```suggestion
You may experience incompatibility issues if you use a system-wide install of `pip` in conjunction with an active `conda` environment.
```

On Windows, the easiest way to find out which `pip` you are using is to invoke `where pip` on the anaconda prompt that you are using.
The `pip` executable that will be called when you invoke it from the command prompt will be at the top and should be located in your conda environment.
You can verify this by checking whether the path contains your currently active environment in it.

> If you see that `pip` is either not on your path or it is being sourced from a different location, double check whether you have activated the correct conda environment.

Reviewer: ```suggestion
If you see that `pip` is either not on your path or it is being sourced from a different location, verify whether you have activated the correct conda environment.
```

You can activate an environment by invoking `conda activate <environment name>`, where `<environment name>` should be replaced with the environment you want to activate.

> If you have already activated the correct environment, but you still don't see `pip` being sourced from there then you can install it into this environment by invoking `conda install pip`.

Reviewer: ```suggestion
If you have already activated the correct environment, but you still do not see `pip` being fetched from it then you can install it into this environment by invoking `conda install pip`.
```

    lk.pytest()

If all tests pass (except for the slow ones which are skipped) then your installation of `pylake` is good to go.

**I tried the installation instructions but conda still won't install pylake**

> If creating a new environment doesn't work then it may be time to uninstall and reinstall conda.

Reviewer: ```suggestion
If creating a new environment does not work then it may be best to uninstall and reinstall conda.
```

> Note that this means you will lose all the environments you have created.

Reviewer: This should be highlighted somehow, lest a user rushes to do the above step unaware of this effect.

Author: done

Please follow these `uninstall instructions`_ to uninstall conda.
After uninstalling, you should be able to reinstall using the regular installation instructions.

**Conda-forge is very slow in China, what can I do?**

Conda-forge can be slow when accessed from China.
This can be resolved by using a Chinese mirror to install Pylake.

> Since there is no mirror for 'conda-forge', Pylake then has to be installed using pip, as outlined below.

Reviewer: ```suggestion
Since there is no mirror for `conda-forge`, Pylake then has to be installed using pip, as outlined below.
```

Note that this means you will lose all the environments you have created.
Please follow these `uninstall instructions`_ to uninstall conda.
After uninstalling, you should be able to reinstall using the regular installation instructions.

**Conda-forge is very slow in China, what can I do?**

Conda-forge can be slow when accessed from China.
In this case, it may be wise to use a Chinese mirror to install Pylake.
Unfortunately, there is no mirror for `conda-forge`.
This means that we cannot use the regular `conda-forge` installation instructions that were outlined above.

> 

Reviewer 2: I'd prefer slightly different wording:

Conda-forge can be slow when accessed from China. This can be resolved by using a Chinese mirror to install Pylake. Since there is no mirror for 'conda-forge', Pylake then has to be installed using pip, as outlined below.

Author: Yes, much better! Amended!

If you normally manage your environments with `pip`, you can just invoke::

    pip install -i https://pypi.tuna.tsinghua.edu.cn/simple lumicks.pylake

> If you use Anaconda, then it is wise to create a new environment for this installation. You can do this as follows::

Reviewer 2: I'd prefer not to use 'wise' in this context:

If you use Anaconda, then it is best to create...

Author: True, that sounds a lot better. Amended!
